hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,223 @@
+    if (!fieldRequired(fieldIdx))
+      Region.clearBit(offset, missingIdx(fieldIdx).toLong)
+    else
+      Code._empty
+  }
+
+  def fieldOffset(structAddress: Long, fieldIdx: Int): Long =
+    structAddress + byteOffsets(fieldIdx)
+
+  def fieldOffset(structAddress: Code[Long], fieldIdx: Int): Code[Long] =
+    structAddress + byteOffsets(fieldIdx)
+
+  def loadField(offset: Long, fieldIdx: Int): Long = {
+    val off = fieldOffset(offset, fieldIdx)
+    types(fieldIdx).fundamentalType match {
+      case _: PArray | _: PBinary => Region.loadAddress(off)
+      case _ => off
+    }
+  }
+
+  def loadField(offset: Code[Long], fieldIdx: Int): Code[Long] =
+    loadField(fieldOffset(offset, fieldIdx), types(fieldIdx))
+
+  private def loadField(fieldOffset: Code[Long], fieldType: PType): Code[Long] = {
+    fieldType.fundamentalType match {
+      case _: PArray | _: PBinary => Region.loadAddress(fieldOffset)
+      case _ => fieldOffset
+    }
+  }
+
+  def deepPointerCopy(mb: MethodBuilder, region: Code[Region], dstStructAddress: Code[Long]): Code[Unit] = {
+    var c: Code[Unit] = Code._empty
+
+    var i = 0
+    while(i < this.size) {
+      val dstFieldType = this.fields(i).typ.fundamentalType
+      if(dstFieldType.containsPointers) {
+        val dstFieldAddress = mb.newField[Long]
+        c = Code(
+          c,
+          this.isFieldDefined(dstStructAddress, i).orEmpty(
+            Code(
+              dstFieldAddress := this.fieldOffset(dstStructAddress, i),
+              dstFieldType match {
+                case t@(_: PBinary | _: PArray) =>
+                  Region.storeAddress(dstFieldAddress, t.copyFromType(mb, region, dstFieldType, Region.loadAddress(dstFieldAddress), forceDeep = true))
+                case t: PCanonicalBaseStruct =>
+                  t.deepPointerCopy(mb, region, dstFieldAddress)
+              }
+            )
+          )
+        )
+      }
+      i += 1
+    }
+
+    c
+  }
+
+  def deepPointerCopy(region: Region, dstStructAddress: Long) {
+    var i = 0
+    while(i < this.size) {
+      val dstFieldType = this.fields(i).typ.fundamentalType
+      if(dstFieldType.containsPointers && this.isFieldDefined(dstStructAddress, i)) {
+        val dstFieldAddress = this.fieldOffset(dstStructAddress, i)
+        dstFieldType match {
+          case t@(_: PBinary | _: PArray) =>
+            Region.storeAddress(dstFieldAddress, t.copyFromType(region, dstFieldType, Region.loadAddress(dstFieldAddress), forceDeep = true))
+          case t: PCanonicalBaseStruct =>
+            t.deepPointerCopy(region, dstFieldAddress)
+        }
+      }
+      i += 1
+    }
+  }
+
+  def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcStructAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+    assert(sourceType.size == this.size)
+
+    if (this == sourceType && !forceDeep)
```

Comment:
So the reason I didn't use type equality was that the semantics of this function allow for requiredenss of srcPType (but not its children) to differ from destType while still allowing for shallow copy, since srcAddress (here srcStructAddress) is necessarily present. I think every place we have this == sourceType, should be reverted to the original form.